pkg/modifycontroller/modify_volume.go

```diff
@@ -0,0 +1,167 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modifycontroller
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-csi/external-resizer/pkg/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	v1 "k8s.io/api/core/v1"
+	storagev1alpha1 "k8s.io/api/storage/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+// The return value bool is only used as a sentinel value when function returns without actually performing modification
+func (ctrl *modifyController) modify(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
+	pvcSpecVacName := pvc.Spec.VolumeAttributesClassName
+	curVacName := pvc.Status.CurrentVolumeAttributesClassName
+	targetVacName := pvc.Status.ModifyVolumeStatus.TargetVolumeAttributesClassName
+
+	if pvcSpecVacName != nil && curVacName != nil && *pvcSpecVacName != *curVacName {
+		if *curVacName == targetVacName {
+			return ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
+		} else {
+			// Check if the PVC is in uncertain State
+			pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+			if err != nil {
+				return pvc, pv, err, false
+			}
+			_, ok := ctrl.uncertainPVCs[pvcKey]
+			if !ok {
+				// PVC is not in uncertain state
+				klog.V(3).InfoS("previous operation on the PVC failed with a final error, retrying")
+				return ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
+			} else {
+				vacObj, _, err := ctrl.volumeAttributesClasses.GetByKey(*pvcSpecVacName)
+				if err != nil {
+					return pvc, pv, err, false
+				}
+				return ctrl.controllerModifyVolumeWithTarget(pvc, pv, vacObj.(*storagev1alpha1.VolumeAttributesClass), pvcSpecVacName)
+			}
+		}
+
+	}
+	// No modification required
+	return pvc, pv, nil, false
+}
+
+// func validateVACAndModifyVolumeWithTarget validate the VAC. The function sets pvc.Status.ModifyVolumeStatus
+// to Pending if VAC does not exist and proceeds to trigger ModifyVolume if VAC exists
+func (ctrl *modifyController) validateVACAndModifyVolumeWithTarget(
+	pvc *v1.PersistentVolumeClaim,
+	pv *v1.PersistentVolume) (*v1.PersistentVolumeClaim, *v1.PersistentVolume, error, bool) {
+	// The controller only triggers ModifyVolume if pvcSpecVacName is not nil nor empty
+	pvcSpecVacName := pvc.Spec.VolumeAttributesClassName
+	// Check if pvcSpecVac is valid and exist
+	vacObj, exists, err := ctrl.volumeAttributesClasses.GetByKey(*pvcSpecVacName)
+	if exists && err == nil {
+		// Mark pvc.Status.ModifyVolumeStatus as in progress
+		pvc, err = ctrl.markControllerModifyVolumeStatus(pvc, v1.PersistentVolumeClaimModifyVolumeInProgress, nil)
```

Comment:
Is this error value ignored?

Reply:
This error is important, because I am not sure if we want to proceed with volume modification if there was an error updating the PVC in api server.

Reply:
Added error handling